src/runtime/audience-action-local-flow.ts

```diff
@@ -161,11 +173,64 @@ export class AudienceActionLocalFlow implements AudienceActionFlow {
   private async initPrompt_() {
     if (this.params_.action === TYPE_REWARDED_AD) {
       await this.initRewardedAdWall_();
+    } else if (this.params_.action === TYPE_NEWSLETTER_SIGNUP) {
+      await this.initNewsletterSignup_();
+    } else {
+      this.renderErrorView_();
+    }
+  }
+
+  private async initNewsletterSignup_() {
+    this.eventManager_.logSwgEvent(
+      AnalyticsEvent.IMPRESSION_BYOP_NEWSLETTER_OPT_IN
+    );
+    const config = await this.getConfig_();
+    const codeSnippet = config?.optInParameters?.codeSnippet;
+
+    const validNewsletterSignupParams =
+      codeSnippet &&
+      config?.optInParameters?.promptPreference ===
+        PREFERENCE_PUBLISHER_PROVIDED_PROMPT;
+
+    if (validNewsletterSignupParams) {
+      this.renderOptinPrompt_(codeSnippet);
     } else {
+      this.eventManager_.logSwgEvent(
+        AnalyticsEvent.EVENT_BYOP_NEWSLETTER_OPT_IN_CONFIG_ERROR
+      );
       this.renderErrorView_();
     }
   }
 
+  private renderOptinPrompt_(codeSnippet: string) {
+    if (!codeSnippet.includes('form')) {
```

Comment:
Rather than: 
1. Running a naive string check for 'form'
2. Setting innerHTML on the prompt
3. Checking for a <form> element

Would it make more sense to validate this once, eg by creating a new div, setting innerHTML, and checking its structure, before modifying prompt_?

Reply:
Done! Checking for a <form> element is required for prompt_ since the result of querySelector() could be null, even if we check the <form> element in another div, so I had to replace prompt_ if we only want to check for 'form' once.